src/backend/cuda/convolve.cpp

```diff
@@ -112,5 +125,311 @@ INSTANTIATE(ushort, float)
+                                           fDims[2], fDims[1], fDims[0]));
+
+    // create convolution descriptor
+    cudnnConvolutionDescriptor_t convolution_descriptor;
+    CUDNN_CHECK(cudnnCreateConvolutionDescriptor(&convolution_descriptor));
+
+    CUDNN_CHECK(cudnnSetConvolution2dDescriptor(
+        convolution_descriptor, padding[1], padding[0], stride[1], stride[0],
+        dilation[1], dilation[0], CUDNN_CONVOLUTION, cudnn_dtype));
+
+    // get output dimensions
+    const int tensorDims = 4;
+    int convolved_output_dim[tensorDims];
+    CUDNN_CHECK(cudnnGetConvolutionNdForwardOutputDim(
+        convolution_descriptor, input_descriptor, filter_descriptor, tensorDims,
+        convolved_output_dim));
+
+    // create output descriptor
+    const int n_out = convolved_output_dim[0];
+    const int c_out = convolved_output_dim[1];
+    const int h_out = convolved_output_dim[2];
+    const int w_out = convolved_output_dim[3];
+
+    cudnnTensorDescriptor_t output_descriptor;
+    CUDNN_CHECK(cudnnCreateTensorDescriptor(&output_descriptor));
+    CUDNN_CHECK(cudnnSetTensor4dDescriptor(output_descriptor, CUDNN_TENSOR_NCHW,
+                                           cudnn_dtype, n_out, c_out, h_out,
+                                           w_out));
+
+    // get convolution algorithm
+    const int memory_limit =
+        0;  // TODO: set to remaining space in memory manager?
+    cudnnConvolutionFwdAlgo_t convolution_algorithm;
+    CUDNN_CHECK(cudnnGetConvolutionForwardAlgorithm(
+        cudnn, input_descriptor, filter_descriptor, convolution_descriptor,
+        output_descriptor, CUDNN_CONVOLUTION_FWD_PREFER_FASTEST, memory_limit,
+        &convolution_algorithm));
+
+    // figure out scratch space memory requirements
+    size_t workspace_bytes;
+    CUDNN_CHECK(cudnnGetConvolutionForwardWorkspaceSize(
+        cudnn, input_descriptor, filter_descriptor, convolution_descriptor,
+        output_descriptor, convolution_algorithm, &workspace_bytes));
+
+    // prepare output array and scratch space
+    dim4 odims(w_out, h_out, c_out, n_out);
+    Array<accT> out = createEmptyArray<accT>(odims);
+
+    auto workspace_buffer = memAlloc<char>(workspace_bytes);
+
+    // perform convolution
+    accT alpha = scalar<accT>(1.0);
+    accT beta  = scalar<accT>(0.0);
+    CUDNN_CHECK(cudnnConvolutionForward(
+        cudnn, &alpha, input_descriptor, cast<accT>(signal).device(),
+        filter_descriptor, filter.device(), convolution_descriptor,
+        convolution_algorithm, (void *)workspace_buffer.get(), workspace_bytes,
+        &beta, output_descriptor, out.device()));
+
+    // destroy all descriptors
+    CUDNN_CHECK(cudnnDestroyTensorDescriptor(input_descriptor));
+    CUDNN_CHECK(cudnnDestroyTensorDescriptor(output_descriptor));
+    CUDNN_CHECK(cudnnDestroyFilterDescriptor(filter_descriptor));
+    CUDNN_CHECK(cudnnDestroyConvolutionDescriptor(convolution_descriptor));
```

Comment:
@syurkevi @umar456 do you think there will be advantage if we cache these descriptors ?

Irrespective of that, we should wrap these cudnn resources to their respective RAII constructs. we already have cudnn.hpp/cpp , we should just place those definitions in there.

Reply:
I think we looked into it and it didn't take a long time to create these. I think it may make a difference if add convolutions to the JIT tree. 

I agree these should be wrapped in a unique_ptr or something.